event-schema/counter_manager.go

```diff
@@ -36,13 +54,41 @@ func getAllFrequencyCounters(schemaHash string) []*FrequencyCounter {
 	return frequencyCounters
 }
 
-func getFrequencyCounter(schemaHash, key string) *FrequencyCounter {
+func getFrequencyCounterBounded(schemaHash, key string, bound int) *FrequencyCounter {
 	schemaVersionCounters, ok := countersCache[schemaHash]
 	if !ok {
 		schemaVersionCounters = make(map[string]*FrequencyCounter)
 		countersCache[schemaHash] = schemaVersionCounters
 	}
```

Comment:
Are those operations thread/concurrent safe?
`countersCache` is used by multiple methods, if any of the calls are concurrent it is going to be an issue

Also `counterCache` is a global variable, it would make more sense to create a struct around it and use methods instead of functions to edit it. I understand this is beyond the scope of this PR.

```go

type FrequencyCounter struct {
    // maybe a mutext
    countersCache map[string]map[string]*FrequencyCounter
}


func (f *FrequencyCounter) func getFrequencyCounterBounded(schemaHash, key string, bound int) *FrequencyCounter {

...
}
```

Reply:
The way it is written originally is to take a very wide lock and then perform operations on all global data structures. This works okay for currently functionality as code is a bit limited in scope but it is bound to create issues as mentioned above given multiple workers will be working on the same data structures.